### ISSUE TITLE
ring attention dpo loss 不能对齐

### ISSUE THREAD

**Issue by xs1997zju** [OpenRLHF/OpenRLHF#564]
您好，感谢如此优秀的框架，我在进行单机8卡， dpo训练时， 当开启ring-attention时， loss与未开启该配置时diff较大， 且ring-attn-size越大， loss震荡幅度越大， 如下图， 蓝色是未开启ring-attn的曲线， 红色是ring-attn-size=2,  灰色是ring-attn-size=8；
<img width="892" alt="image" src="https://github.com/user-attachments/assets/bc1d0de6-7320-4388-ae01-c2d9544504fc">

**Comment by xs1997zju** [comment]
是因为开启ring-attn后， backward出的梯度与未开启ring有较大diff？

**Comment by hijkzzz** [comment]
@zhuzilin

**Comment by xs1997zju** [comment]
@zhuzilin 
黄色曲线：train_bs=8, micro=1, ring-size=1
蓝色曲线：train_bs=8, micro=1, ring-size=8
<img width="1579" alt="image" src="https://github.com/user-attachments/assets/7680cae4-63cd-4ddd-a154-73b84d4af12b">

**Comment by xs1997zju** [comment]
@zhuzilin  train_bs=8, micro=8, ring-size=1, 这个设置在目前的代码中设置是报错的， 我看代码， train_bs是global batch size, 关系是train_bs = micro * world_size,  train_bs=8, micro=8这样的设置是非法的， micro=8， train_bs至少是64 （单机八卡）

**Comment by xs1997zju** [comment]
团队是不是能够再review下ring-attnention在openrlhf的适配逻辑？ 长文本的dpo， ring-attention在显存方面还是很有优势的

**Comment by xs1997zju** [comment]
train_bs=8, micro=8, ring-size=8,这个设置也是报错的
<img width="1134" alt="image" src="https://github.com/user-attachments/assets/8210fff9-6425-426b-ada5-b5aac57d531f">

**Comment by xs1997zju** [comment]
@zhuzilin train_bs=8, micro=8, ring-size=8, 这个设置也是报错的
<img width="896" alt="image" src="https://github.com/user-attachments/assets/86627529-5994-41a0-a80e-cf5cbe46a1d9">

**Comment by zhuzilin** [comment]
grad_acc < 0 的原因应该是您的 num_gpu / ring_size * micro_bs > train_bs 了.... 相当于 train bs 设置的太小了....

**Comment by xs1997zju** [comment]
@hijkzzz 但是你这里的意思， 不是 ring attn 之后，将 micro_train_batch_size 乘 ring_attn_size 来进行训练吗， 即train_bs = micro * ring-size ?

**Comment by xs1997zju** [comment]
@zhuzilin 但是你这里的意思， 不是 ring attn 之后，将 micro_train_batch_size 乘 ring_attn_size 来进行训练吗， 即train_bs = micro * ring-size ?

**Comment by Freder-chen** [comment]
> train_bs=8, micro=1, ring-size=8

这里要对齐，设置的实验应该是：train_bs=64, micro=1, ring-size=8吧。但似乎会与dataset的random seed有关，会有细微差别。

**Comment by zhuzilin** [comment]
> 但是你这里的意思， 不是 ring attn 之后，将 micro_train_batch_size 乘 ring_attn_size 来进行训练吗， 即train_bs = micro * ring-size ?

不是... 我的意思是 `mbs = old_mbs * ring_size`

**Comment by xs1997zju** [comment]
> > 但是你这里的意思， 不是 ring attn 之后，将 micro_train_batch_size 乘 ring_attn_size 来进行训练吗， 即train_bs = micro * ring-size ?
> 
> 不是... 我的意思是 `mbs = old_mbs * ring_size`

@zhuzilin 这样设置， ring降低显存的作用似乎就不明显了， 以单机8卡为例， 假设old_mbs=1,   ring_size=8,  mbs=8，  开启ring后，所有卡看到的data parallel group是相同的， mbs=8的话， pack后再ring切分， 相当于每个卡上见到一个全长度的样本， 而不是1/8长度的样本， 这样在单条样本max-len比较大的情况下， ring没有起到降低显存的作用？

**Comment by Freder-chen** [comment]
> 嗯.... 我的意思是说，如果您希望的是对齐 loss，就需要把开 ring attn 和不开的数据配置对齐才能正确对比。在这个情况下，需要把 mbs 的倍数加上。这样得到的差别才是 ring attn 带来的差别，不然 loss 的波动应该是因为 packing samples 每个 mbs 不等长导致的。

您好，今天刚好有相关疑问，歪楼问一下。

如果我想对齐Total Batch Size，对于原始配置train_bs=8, micro=1, ring-size=1
在打开ring-attn后，是不是应该配置train_bs=64, micro=1, ring-size=8

**Comment by zhuzilin** [comment]
应该是要配置成 train_bs=8, micro=8, ring-size=8。
可以理解成，之前是 1 张卡（ring-size=1）跑 1 条sample（micro=1），如果要对齐，那么对应的应该是 8 张卡（ring-size=8）跑 8 条 sample（micro=8）

**Comment by xs1997zju** [comment]
@zhuzilin 感谢，我想可能应该是显示的原因， tensorboard记录的是单步的loss， 我这里可能搞混概念了；
理论上， 当global-bs一致的时候， 不论pack还是不pack， global loss 以及 global 梯度应该是一致的， 不知道我这个理解对不对哈

**Comment by xs1997zju** [comment]
因为我看就是dpo的loss，即使是packing后， 我看最后get_prob还是按unpack之后的情况来计算logits-sum和最终的损失：

**Comment by xs1997zju** [comment]
对于单机单卡上的，mbs=1而言， pack或者不pack， loss应该是大致一样的

**Comment by xs1997zju** [comment]
简单修改了下loss_mean的逻辑：
<img width="719" alt="image" src="https://github.com/user-attachments/assets/257dd6de-47c8-40f1-aad2-4717665536ff">

<img width="711" alt="image" src="https://github.com/user-attachments/assets/1e7868e2-a512-4487-9e15-bafe95f02947">

从loss_mean的角度看， train_bs=8, mirco_bs=1, ring=1, 和 train_bs=8, mirco_bs=1, ring=8的曲线基本一致：
<img width="711" alt="image" src="https://github.com/user-attachments/assets/494b09d2-ce48-4ca5-922f-b5f5319e2df4">

**Comment by xs1997zju** [comment]
> grad_acc < 0 的原因应该是您的 num_gpu / ring_size * micro_bs > train_bs 了.... 相当于 train bs 设置的太小了....

@zhuzilin 一点小建议，感觉代码里计算梯度累积的逻辑可以稍稍修改下：
<img width="811" alt="image" src="https://github.com/user-attachments/assets/649125e8-d662-41a3-8097-bb2fee09ae50">
比如说单机八卡， 在开ring=8的时候， global-size是可以设置为1的， 8张卡对同一个sample进行序列并行， 但是按现在的写法， 由于world_size=8， 前面除法整除会出现0， 再乘ring-size也还是0， 可以把ring-size的乘放到最前面
这样， global batchsize-1,   micro-1, ring-8, 类似的配置也就都合法了

**Comment by zhuzilin** [comment]
赞，我去改一下~
